### ISSUE TITLE
check file syntax for legacy fonts. 

### ISSUE THREAD

**Issue by u-fischer** [u-fischer/luaotfload#11]
The documentation claims that a legacy font can be used with `file:` (section 5.4.1)
But 

~~~~
\input luaotfload.sty
\font \test = {file:cmr10.tfm} at 10pt  % example from the doc
\test A
\bye
~~~~

fails with 

~~~~
xmf-var/fonts/map/pdftex/updmap/pdftex.map}])kpathsea:make_tex: Invalid filename `file:cmr10.tfm', contains ':'

! error:  (file file:cmr10.tfm) (type 3): font file:cmr10.tfm at 600 not found
!  ==> Fatal error occurred, no output PDF file produced!Drücken Sie eine beliebige Taste . . .
~~~~

See also https://github.com/lualatex/luaotfload/issues/408.

**Comment by zauguin** [comment]
This is caused by the `rewrite_fontname` from `luaotfload-auxiliary`.
So an easy fix would be disabling rewriting with

    \directlua{luaotfload.aux.stop_rewrite_fontname()}
but `patch_font` callbacks should not be called for TFM fonts in the first place.
That is what `patch_font_unsafe` is for AFAICT.

**Comment by u-fischer** [comment]
I agree that tfm fonts shouldn't perhaps use the patch_font callback. The comment "or else the fontspec capheight callback will operate on tfm fonts." is quite clear (but we should check if the callback is still there ...).  So I merged. Let's hope that nobody used the callback on a tfm font somewhere -- the documentation doesn't mention the alternative `patch_font_unsafe`. 

Let's also hope that the test for format not unknown will stay reliable. The fontdump of a tfm font, shows `["format"]="unknown",` as main entry but `["format"]="type1",` in the properties -- this looks as if something is not correctly duplicated (all this duplicates are unnerving anyway, I wonder if there are really packages around that still rely on old table organisation). 

Beside this the organizsation and use of two patch commands looks a bit messy.   There is no clear description which tables a font need to be "patchable".  Quite a number seem to have been added quite adhoc and never reviewed later. 

E.g. Imho the patch fixup_fontdata in luaotfload-auxiliary isn't needed  (anymore). If I disable it tfm fonts nevertheless have a resource table and the example in https://github.com/lualatex/luaotfload/issues/253 compiles without error.

**Comment by u-fischer** [comment]
@zauguin  I merged the pull request and while running a number of tests I also run this example:

~~~~
\documentclass{article}

\usepackage{luacode}

\begin{luacode}

local testpatch = function (fontdata)
  texio.write_nl("patch_font") 
end

local testpatchunsafe = function (fontdata)
  texio.write_nl("patch_font_unsafe")
 end


luatexbase.add_to_callback
 (
  "luaotfload.patch_font",
   testpatch,
  "test"
 )

luatexbase.add_to_callback
 (
  "luaotfload.patch_font_unsafe",
   testpatchunsafe,
  "test"
 )

 \end{luacode}
\begin{document}
\font\test={name:texgyretermes}\test



\font\test=cmr10 \test

blblb

\font\test={name:texgyretermes}\test

blbl

\end{document}
~~~~

This outputs

~~~~
patch_font
patch_font_unsafe
patch_font_unsafe

~~~~

The first two are as expected. But what about the third? (it is not related to the patch, one got the last also previously).

**Comment by zauguin** [comment]
@u-fischer The third one is the reason we check for `type(...)=='table'`: The font is already loaded and has an id, so instead of a new font table the ID of the already existing font is returned.

In this case `patch_font_unsafe` is called because the font is most likely already patched and we only have an ID instead of a modifiable table, so patching would be literally "unsafe".

**Comment by u-fischer** [comment]
@zauguin  yes, but what I meant (was a bit late yesterday): do we want a third code path here?

As an example I can add the `dump_font` function to `patch_font_unsafe` (https://github.com/u-fischer/lua-font-pond/blob/master/plain/dump-font.tex) and at the `\font\test=cmr10` it will create a fontdump of cmr10 without problem. But at the second call of `texgyretermes` one gets an (expected) lua error `bad argument #1 to 'next' (table expected, got number)`. 

Imho we should do

* if table and encodingbytes ==  2  -> `patch_font`
* if table and encodingbytes ==  0 _> `patch_font_unsafe` (perhaps renamed ...)
* if not table: don't patch ...

Or do you see any case where in the last case one would want to apply a patch? Then we should perhaps call it `patch_font_very_unsafe_as_no_table` or something like this ;-).

**Comment by zauguin** [comment]
@u-fischer Sorry for the late response, I got distracted.

My problem with this is that I do not think patching TFM fonts is advisable in any way.
There might be slightly more use-cases for this than "patching" a number, but most attempts to patch such a font will probably break.
Especially the not existing `index` key in the characters prohibit a lot of patching opportunities.

So there are two cases where the current `patch_font_unsafe` is useful:

- You really now what you are doing and you really want to patch a TFM font. Then you can just check yourself if you really got a table.
- You use the `patch_font(_unsafe)` callbacks for reporting/debugging/... While `dump_font` falls into this category and needs a table, generally I thing here calling the callback for font ids too is an advantage: For example when tracking font definitions, this gives you the choice if you want to ignore redefinitions or if you want to track everything.

So IMHO not calling the callback for ID's would make some rare cases easier, but other rare cases much more complicated.
Also this leads to a much clearer error message if someones tries to misuse `patch_font_unsafe` by just throwing a `patch_font` callback in without checking if it is safe.

tl;dr: I do not think we should change this, but *if* we change this I would at least want some form of `patch_font_very_unsafe`.

**Comment by u-fischer** [comment]
OK. So we have `patch_font` for "open type" fonts (table and encodingbyte=2) and `patch_font_unsafe` for the rest. And everyone who wants to use `patch_font_unsafe` (like me with my fontdump ;-)) should *inside* the patch function test, if the argument is actually a font table or not. That sounds sensible.

**Comment by u-fischer** [comment]
v 2.92 pushed to ctan so closing the issue.
